File: src/main/java/komu/blunt/types/checker/PatternTypeCheckResult.java
package komu.blunt.types.checker;

import komu.blunt.types.Predicate;

import java.util.Collections;
import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;

final class PatternTypeCheckResult<T> {

    final List<Predicate> predicates;
    final Assumptions as;
    final T value;

    PatternTypeCheckResult(List<Predicate> predicates, Assumptions as, T type) {
        this.predicates = checkNotNull(predicates);
        this.as = checkNotNull(as);
        this.value = checkNotNull(type);
    }

    PatternTypeCheckResult(Assumptions as, T type) {
        this.predicates = Collections.emptyList();
        this.as = checkNotNull(as);
        this.value = checkNotNull(type);
    }

    @Override
    public String toString() {
        return predicates + "/" + as + "/" + value;
    }
}

File: src/main/kotlin/komu/blunt/types/checker/PatternTypeCheckResult.kt
package komu.blunt.types.checker

import komu.blunt.types.Predicate

class PatternTypeCheckResult<out T> constructor(val value: T, val ass: Assumptions = Assumptions.empty, val predicates: List<Predicate> = emptyList()) {

    // TODO this(ass: Assumptions, value: T): this(Collections.emptyList<Predicate>().sure(), ass, value) { }

    override fun toString() = "$predicates/$ass/$value"
}
